src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityAssemblyLine.java

```diff
@@ -60,4 +70,127 @@ public ICubeRenderer getBaseTexture(IMultiblockPart sourcePart) {
     protected IBlockState getCasingState() {
         return MetaBlocks.METAL_CASING.getState(BlockMetalCasing.MetalCasingType.STEEL_SOLID);
     }
+
+    private int beamCount;
+    @SideOnly(Side.CLIENT)
+    private GTLaserBeamParticle[][] beamParticles;
+
+    @Override
+    public void update() {
+        super.update();
+        if(ConfigHolder.client.shader.assemblyLineParticles) {
+            if (getRecipeMapWorkable().isWorking() || getRecipeMapWorkable().isActive()) {
+                int maxBeams = getAbilities(MultiblockAbility.IMPORT_ITEMS).size() + 1;
+                int maxProgress = getRecipeMapWorkable().getMaxProgress();
+
+                // Each beam should be visible for an equal amount of time, which is derived from the maximum number of
+                // beams and the maximum progress in the recipe.
+                int beamTime = Math.max(1, maxProgress / maxBeams);
+
+                int currentBeamCount = Math.min(maxBeams, getRecipeMapWorkable().getProgress() / beamTime);
+
+                if (currentBeamCount != beamCount) {
+                    beamCount = currentBeamCount;
+                    writeCustomData(GregtechDataCodes.UPDATE_PARTICLE, this::writeParticles);
+                }
+                return;
```

Comment:
Instead of returning, it may be better to use an `else if (beamCount != 0)`, so any future additions to `update()` are always called independently of the lasers.